src/crypto/serde_impl.rs

```diff
@@ -80,3 +80,105 @@ pub mod projective_vec {
         Ok(wrap_vec.into_iter().map(|CurveWrap(c, _)| c).collect())
     }
 }
+
+/// Serialization and deserialization of vectors of field elements.
+pub mod field_vec {
+    use std::borrow::Borrow;
+    use std::marker::PhantomData;
+
+    use pairing::{PrimeField, PrimeFieldRepr};
+    use serde::de::Error as DeserializeError;
+    use serde::ser::Error as SerializeError;
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+    /// A wrapper type to facilitate serialization and deserialization of field elements.
+    pub struct FieldWrap<F, B>(B, PhantomData<F>);
+
+    impl<F, B> FieldWrap<F, B> {
+        pub fn new(f: B) -> Self {
+            FieldWrap(f, PhantomData)
+        }
+    }
+
+    impl<F> FieldWrap<F, F> {
+        pub fn into_inner(self) -> F {
+            self.0
+        }
+    }
+
+    impl<F: PrimeField, B: Borrow<F>> Serialize for FieldWrap<F, B> {
+        fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+            let mut bytes = Vec::new();
+            self.0
+                .borrow()
+                .into_repr()
+                .write_be(&mut bytes)
+                .map_err(|_| S::Error::custom("failed to write bytes"))?;
+            bytes.serialize(s)
+        }
+    }
+
+    impl<'de, F: PrimeField> Deserialize<'de> for FieldWrap<F, F> {
+        fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+            let bytes: Vec<u8> = Deserialize::deserialize(d)?;
+            let mut repr = F::zero().into_repr();
+            repr.read_be(&bytes[..])
+                .map_err(|_| D::Error::custom("failed to write bytes"))?;
+            Ok(FieldWrap::new(F::from_repr(repr).map_err(|_| {
+                D::Error::custom("invalid field element representation")
+            })?))
+        }
+    }
+
+    pub fn serialize<S, F>(vec: &[F], s: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        F: PrimeField,
+    {
+        let wrap_vec: Vec<FieldWrap<F, &F>> = vec.iter().map(FieldWrap::new).collect();
+        wrap_vec.serialize(s)
+    }
+
+    pub fn deserialize<'de, D, F>(d: D) -> Result<Vec<F>, D::Error>
+    where
+        D: Deserializer<'de>,
+        F: PrimeField,
+    {
+        let wrap_vec = <Vec<FieldWrap<F, F>>>::deserialize(d)?;
+        Ok(wrap_vec.into_iter().map(|FieldWrap(f, _)| f).collect())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use bincode;
+    use pairing::bls12_381::Bls12;
+    use pairing::Engine;
+    use rand::{self, Rng};
+
+    #[derive(Debug, Serialize, Deserialize)]
+    pub struct Vecs<E: Engine> {
```

Comment:
Should the tests be kept with the `Engine` parameter? Since you are using `Bls12` anyway, maybe the tests can be simplified too?

Reply:
Yes, I kept the `serde_impl` module generic for now. If https://github.com/ebfull/pairing/pull/87 gets merged, this module can be removed anyway. Otherwise I'll update it next week in a separate PR.